Fix function inlining when only interface is available (fixes #397)

reuterbal opened the pull request:
This adds the missing test for `InlineSubstitutionMapper`, expanded to four different cases:

1. Inline call provided as a module import, module available
2. Inline call provided as a module import, module not available
3. Inline call provided via an interface block
4. Inline call provided via an interface that is imported via c preprocessor (i.e., no interface available)

This includes a small fix to the mapper for the third case, where procedure information is available but that referes only to the interface, i.e., no body for inlining is present.

Let's see what codecov says...

mlange05 reviewed (approved):
Ok, not a fan but since this only increases test coverage, it's ready to go.

As a side-note, I think we need a separate re-write of `transformation.inline` module, as the new capabilities from PR #378 basically voided the previous separation by callee-type (which was an artifact of only allowing single-line substitutions). The module now also has cyclic references and needs an general overhaul, I think, but that beyond this PR.

reuterbal commented on loki/transformations/inline/tests/test_functions.py:
```diff
@@ -239,3 +240,91 @@ def test_inline_statement_functions(frontend, stmt_decls):
+    assignments = FindNodes(ir.Assignment).visit(routine.body)
+
+    if provide_myfunc in ('import', 'module', 'intfb'):
+        assert len(inline_calls) == 0  # MYFUNC(arr) is misclassified as array subscript or fully inlined
```
It's not a bug, it's a lack of information. There are two cases here, which could maybe be made more explicit in the test:
```
    if provide_myfunc in ('import, 'intfb'):  # We only have the use statement or nothing but the C-include
        assert len(inline_calls) == 0  # MYFUNC(arr) is misclassified as array subscript
    elif provide_myfunc == 'module':  # We have the full function definition
        assert len(inline_calls) == 0  # MYFUNC(arr) is eliminated by inlining
```
In the first case we have that symbol but don't have the interprocedural annotation that explains what this symbol is. Which brings us to the ambiguity of fortran that does not syntactically allow to distinguish this call from an array subscript and fparser classifies it as the latter.
In the second case we eliminate MYFUNC.

reuterbal replied:
Just to add, it's independent from the statement function - the same is true in any other expression

reuterbal commented on loki/transformations/inline/tests/test_functions.py:
```diff
@@ -239,3 +240,91 @@ def test_inline_statement_functions(frontend, stmt_decls):
+    assert assignments[0].lhs  == 'ret'
+    assert assignments[1].lhs  == 'ret2'
+    if provide_myfunc == 'module':
+        assert assignments[0].rhs  ==  "arr + arr + 1.0 + arr*2.0 + arr*2.0"
```
No, the behaviour difference comes again from the lack of information. Inlining always works the same if you have the definition of the function body available, and fails when you don't have it

In fact, let me add a fifth case `provide_myfunc == 'subroutine'`, where we have the intfb and provide the definition separately as a free function (not embedded in a module). Providing this as a definition should also allow the inlining to be effective

reuterbal commented:
Thanks for the review and the helpful comments. I have added a few more explanations to the test and expanded this to include the case that the function is provided via the definitions. In that situation we should also have all the information available to inline the body.

It turns out that we don't use this information in the fparser frontend, though, so I added a small fix for this.
